### PR TITLE
File IO improvements

### PR DESCRIPTION
**Pull request by smavros**
addressees #31 #193 and #312

### CODE REVIEW

**Comment by GPMueller** [comment]
This is also related to #166, #220 and #260.

**Comment by GPMueller** [comment]
Note also #234 - this could maybe be fixed in the update of the `Filter_File_Handle`.

**Comment by GPMueller** [comment]
We also need to improve the Chain IO API functions, see https://github.com/spirit-code/spirit/blob/develop/core/include/Spirit/IO.h#L25 they should look something like this:
```C++
DLLEXPORT int  IO_N_Images_In_File(State * state, const char * file, int format=IO_Fileformat_Regular);
DLLEXPORT void IO_Chain_Read(State * state, const char * file, int format=IO_Fileformat_Regular, int idx_starting_image=-1, int idx_chain=-1);
DLLEXPORT void IO_Chain_Write(State * state, const char * file, int format=IO_Fileformat_Regular, int idx_chain=-1);
DLLEXPORT void IO_Chain_Append(State * state, const char * file, int format=IO_Fileformat_Regular, int idx_starting_image=-1, int idx_chain=-1);
```

To this end, we need to improve how the API uses functions like Dataparser `Read_Spin_Configuration_Chain function, see https://github.com/spirit-code/spirit/blob/develop/core/src/io/Dataparser.cpp#L124
Probably it would be best if the cases like "more images in file than in current chain" were handled in the API instead of the Dataparser functions. For the API to check this kind of thing it makes sense to write a new API function like `IO_N_Images_In_File`

Lastly, `Read_Spin_Configuration_Chain` should make use of `Read_Spin_Configuration` to avoid code duplication.

**Comment by smavros** [comment]
Tests should be added for the IO of different filetypes. TDD can be done with simple trivial tests. Related to #300 #297

File: core/include/Spirit/Exception.h [hunk 1]
```diff
@@ -5,4 +5,5 @@
 #define Exception_Simulated_domain_too_small 3
 #define Exception_Not_Implemented            4
 #define Exception_Non_existing_Image         5
-#define Exception_Non_existing_Chain         6
\ No newline at end of file
+#define Exception_Non_existing_Chain         6
+#define Exception_File_reading_error         7
```
Comment (GPMueller) [inline core/include/Spirit/Exception.h]: This sounds a little too similar to `Exception_File_not_Found`. Maybe `Exception_Bad_File_Content` would be better?

File: core/include/io/Filter_File_Handle.hpp [hunk 2]
```diff
@@ -19,16 +20,19 @@ namespace IO
...
 	public:
+        IO::VF_FileFormat ff;
+        std::string filename;
+        std::unique_ptr<std::ifstream> myfile;
 		std::istringstream iss;
+        
 		// Constructs a Filter_File_Handle with string filename
-		Filter_File_Handle(const std::string& s);
+		Filter_File_Handle( const std::string& s, 
+                            IO::VF_FileFormat format = VF_FileFormat::SPIRIT_GENERAL );
```
Comment (GPMueller) [inline core/include/io/Filter_File_Handle.hpp]: Should the `Filter_File_Handle` really care about the format of the file? I suppose it should be more abstract and simply be created with appropriate comment strings etc.

File: core/src/io/Configparser.cpp [hunk 3]
```diff
@@ -397,29 +397,29 @@ namespace IO
 					IO::Filter_File_Handle myfile(configFile);
 
 					// N_a
-					myfile.Read_Single(na_left, "pin_na_left", false);
-					myfile.Read_Single(na_right, "pin_na_right", false);
-					myfile.Read_Single(na, "pin_na ", false);
+					myfile.Read_Single(na_left, "pin_na_left", SPIRIT_GENERAL, false);
```
Comment (GPMueller) [inline core/src/io/Configparser.cpp]: Why did you add this specifically to the `Read_Single` reading the pinning values?
Also, you currently create a `Filter_File_Handle` with a file format already, so passing it another should be unnecessary. And when the `Filter_File_Handle` instance used here (`myfile`) is configured correctly with comment strings, this read should work inherently without passing extra info into the function.

File: core/src/io/Filter_File_Handle.cpp [hunk 4]
```diff
@@ -10,42 +10,61 @@
...
-		}
+        this->ff = format;
+        this->dump = "";
+        this->line = "";
+        this->found = std::string::npos;
+        this->myfile = std::unique_ptr<std::ifstream>( new std::ifstream( filename,
+                                                       std::ios::in | std::ios::binary ) );
+        
+        // set the comment tag
+        if( this->ff == VF_FileFormat::OVF )
```
Comment (GPMueller) [inline core/src/io/Filter_File_Handle.cpp]: This could be set in the Dataparser function which calls this Class and passed into the constructor as a string, right? That would be more general than giving this class a file format.

File: core/src/io/Configparser.cpp (hunk 3 above)
Comment (smavros) [reply core/src/io/Configparser.cpp]: This is wrong. It was left like that when at first I had the `IO::VF_FileFormat` as argument of `Read_Single()`function. Will be fixed with the next commit.

File: core/include/Spirit/Exception.h (hunk 1 above)
Comment (smavros) [reply core/include/Spirit/Exception.h]: You are right. Will be renamed with next commit.

File: core/include/io/Filter_File_Handle.hpp [hunk 5]
```diff
@@ -43,32 +47,45 @@ namespace IO
...
+        // Read a string into var
+        void Read_String( std::string& var, const std::string name, bool log_notfound = true );
+        // Count the words of a string
+        int Count_Words( const std::string& str );
+        // get name 
+        
 		// Reads a single variable into var, with optional logging in case of failure.
-		template <typename T> void Read_Single(T & var, const std::string name, bool log_notfound = true)
+		template <typename T> void Read_Single( T & var, const std::string name,  
+                                                bool log_notfound = true )
```
Comment (smavros) [inline core/include/io/Filter_File_Handle.hpp]: Would it be more useful if we had another argument like `bool critical = false` in case that `Read_Single()` cannot find a certain `var`, critical for the data read in? In case that the `var` is not found a `Warning` is logged and the default value is used but is that enough to inform the user for the mishap? Sometimes the default value of `var` might not be meaningful. Same holds for the `Filter_File_Handle::Read_String()`

**Comment by smavros** [comment]
Related to #336

File: core/include/io/Filter_File_Handle.hpp (hunk 5 above)
Comment (GPMueller) [reply core/include/io/Filter_File_Handle.hpp]: Actually, it might be better design to simply make this function return a bool. The caller can then decide whether to log a not-found or to e.g. `throw`.

**Comment by GPMueller** [comment]
Concerning CSV vectorfield data it may be worthwile to consider https://github.com/ben-strasser/fast-cpp-csv-parser (see also issue #74).

**Comment by GPMueller** [comment]
Note also issue #34: check if the write functions log appropriate error messages when a file could not be created (e.g. when a folder is non-existent).

**Comment by smavros** [comment]
Many changes must be made so I am making a comprehensive task list (will be extended):

- [ ] add `const char* tag` to API `IO_*_Write()` functions.
- [ ] decide if the `IO_*_Read` functions should have specific `OVF` file format like `OVF_bin8`. Describe the corresponding implementation.
- [ ] implement the `SPIRIT` file format CSV writer and parser.
- [ ] implement the `OVF_bin4` and `OVF_text` writer and parser.
- [ ] make sure that conversion between `IEEE 754 floating types` (for `OVF` IO) 
- [ ] test with a `2x3x4` system without symmetries for `IO` validation.
- [ ] include the `API` changes in the documentation.
- [ ] implement `IO API` functions using `API` to extend the system. 
- [ ] `mu_s` for a unit cell only #297
- [ ] creation of `output_folder` in case of build with C++17
- [ ] `Datawriter.cpp` functions must have arguments `... const vectrofield& vf, const Data::geometry& g ...`

File: core/src/engine/Method_GNEB.cpp [hunk 6]
```diff
@@ -274,26 +274,31 @@ namespace Engine
...
-				preEnergiesFile = this->parameters->output_folder + "/" + starttime + "_Chain_Energies";
-			}
-			else
-			{
-				preChainFile = this->parameters->output_folder + "/Chain";
-				preEnergiesFile = this->parameters->output_folder + "/Chain_Energies";
-			}
+            std::string fileTag;
+            
+            if (this->systems[0]->llg_parameters->output_file_tag == "<time>")
```
Comment (GPMueller) [inline core/src/engine/Method_GNEB.cpp]: As the tag has been abstracted as a parameter across all `Method`s, it would make sense to get this from a function.
However, it's important to use the child class `Parameters` member, (i.e. `gneb_parameters` in the GNEB case)!
At least the latter should be fixed in this PR.

File: core/src/io/Datawriter.cpp [hunk 7]
```diff
@@ -54,10 +58,11 @@ namespace IO
 		if (readability_toggle) header = separator + line + separator;
 		else header = line;
 		if (!readability_toggle) std::replace( header.begin(), header.end(), '|', ' ');
-		String_to_File(header, fileName);
```
Comment (GPMueller) [inline core/src/io/Datawriter.cpp]: All the formatting changes (indentation, where you break lines etc... and now this kind of change... You do too much of this at once! This PR is very hard to keep track of, as you pile a lot of these aesthetic changes on top of the refactoring. Please try to keep this to a minimum and rather create another PR later, where you only do clean-up.
